Register instruments/language in their own compilation units to fix the sbt build issues

JaroslavTulach opened the pull request:
[ci no changelog needed]

### Pull Request Description

New plan to [fix the `sbt` build](https://www.pivotaltracker.com/n/projects/2539304/stories/182209126) and its annoying:
```
log.error(
          "Truffle Instrumentation is not up to date, " +
          "which will lead to runtime errors\n" +
          "Fixes have been applied to ensure consistent Instrumentation state, " +
          "but compilation has to be triggered again.\n" +
          "Please re-run the previous command.\n" +
          "(If this for some reason fails, " +
          s"please do a clean build of the $projectName project)"
        )
```
When it is hard to fix `sbt` incremental compilation, let's restructure our project sources so that each `@TruffleInstrument` and `@TruffleLanguage` registration is in individual compilation unit. Each such unit is either going to be compiled or not going to be compiled as a batch - that will eliminate the `sbt` incremental compilation issues without addressing them in `sbt` itself. 

fa2cf6a33ec4a5b2e3370e1b22c2b5f712286a75 is the first step - it introduces `IdExecutionService` and moves all the `IdExecutionInstrument` API up to that interface. The rest of the `runtime` project then depends only on `IdExecutionService`. Such refactoring allows us to move the `IdExecutionInstrument` out of `runtime` project into independent compilation unit.

### Checklist

- [x]  Move `IdExecutionInstrument` into own source directory `enso/engine/runtime-id-execution/src/main/java/.../IdExecutionInstrument.java` (or else where) and convince `sbt` to compile it (after compiling current `runtime` sources) and then package it into `runtime.jar` using `concat` strategy for `META-INF/services/*` files. 
- [x] Repeat such process with other instruments and languages in the `runtime` project.

JaroslavTulach committed:
Extracting IdExecutionService API from the instrument implementation

radeusgd commented on engine/runtime/src/main/java/org/enso/interpreter/service/ExecutionService.java:
```diff
@@ -171,7 +171,7 @@ public void execute(
     } finally {
       context.getThreadManager().leave(p);
-      listener.dispose();
+        listener.dispose();
```
Why this additional indent?

JaroslavTulach replied:
Applied `javafmt` and produced: [f9cc871](https://github.com/enso-org/enso/pull/3509/commits/f9cc8711512c86f6f00f71b58734b6b6c4675e52)

radeusgd commented on engine/runtime/src/main/java/org/enso/interpreter/instrument/IdExecutionService.java:
```diff
@@ -0,0 +1,264 @@
+    private final QualifiedName typeName;
+    private final String functionName;
+
+    /**
+     * Creates a new instance of this class.
+     *
+     * @param call the function call.
+     */
```
This is unrelated to this PR but I'm wondering if we should keep such comments as they do not really bring any new information.

JaroslavTulach replied:
[f9cc871](https://github.com/enso-org/enso/pull/3509/commits/f9cc8711512c86f6f00f71b58734b6b6c4675e52)

radeusgd reviewed (approved):
Awesome! Very interesting approach to this, cannot wait to see how it helps.

Regarding the "Next Steps" - is there any reason to move the `IdExecutionService` into a separate sub-project in a separate PR instead of performing this whole change within this single PR?

4e6 commented:
> Repeat such process with other instruments and languages in the runtime project.

That sounds like quite a bit of work:sweat_smile:. Wouldn't it be easier to completely disable incremental compilation for runtime project?

JaroslavTulach commented:
> > Repeat such process with other instruments and languages in the runtime project.
> That sounds like quite a bit of worksweat_smile. 

It is four time (two languages, two instruments) more work than in this PR. It is manual work, almost refactoring. 

> Wouldn't it be easier to completely disable incremental compilation for runtime project?

That's what @hubertp was trying to do last month. There doesn't seem to be a way to disable incremental compilation easily. `sbt` tries to be too clever...

> Regarding the "Next Steps

I am sorry for wrong words, I also think these _"Next Steps"_ shall be fixed in this PR. I changed the section title to _"Checklist"_ and added two checkboxes.

4e6 commented on .gitignore:
```diff
@@ -147,3 +147,4 @@ test-results
 
 *.ir
 *.meta
+/nbproject/private/
```
This gitignore should only contain entries related to the project. For files related to your machine please use global gitignore https://gist.github.com/subfuzion/db7f57fff2fb6998a16c

JaroslavTulach replied:
Removed in [d63c0fb](https://github.com/enso-org/enso/pull/3509/commits/d63c0fb1e9d6b34386a4e8d45b6dbd2d13b01c1e) - I am sorry for this accidental change. I didn't do it - my IDE did it behind the scene!

hubertp committed:
Towards instrumentation services split

Split registration of services (now only for `IDExecutionService`) to a
separate project, that way dealing with separate compilation issues in
Truffle's registration DSL.
Added `runtime-instrument` project which now deals with assembling all
classes and resources and most importantly concatenates
`META-INF/services` from individual services.

Unfortunately all instrumentation tests had to be moved as well because
they depend on the moved services.

TBC moving the remaining services and languages registration to
individual projects.

hubertp commented:
> > Repeat such process with other instruments and languages in the runtime project.
> 
> That sounds like quite a bit of worksweat_smile. Wouldn't it be easier to completely disable incremental compilation for runtime project?

Unlikely. Especially because we would want to disable it only for Java sources since by disabling incremental compilation for Scala we would shoot ourselves in the foot.

hubertp committed:
Move remaining Truffle instruments

JaroslavTulach commented on build.sbt:
```diff
@@ -1163,7 +1167,7 @@ lazy val runtime = (project in file("engine/runtime"))
     version := ensoVersion,
     commands += WithDebugCommand.withDebug,
-    cleanInstruments := FixInstrumentsGeneration.cleanInstruments.value,
+    //cleanInstruments := FixInstrumentsGeneration.cleanInstruments.value,
```
If everything works, we can delete this line and also remove the `FixInstrumentsGeneration` class.

hubertp replied:
Yes, this is still WIP

JaroslavTulach commented on build.sbt:
```diff
@@ -1287,6 +1273,133 @@ lazy val runtime = (project in file("engine/runtime"))
+  )
+  .dependsOn(runtime)
+
+lazy val `runtime-instrument`  = (project in file("engine/runtime-instrument"))
```
Possibly name `runtime-with-instruments` would more exactly describe what this project does.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/instrument/ReplExecutionEventNode.java:
```diff
@@ -0,0 +1,44 @@
+
+import java.util.Map;
+
+public interface ReplExecutionEventNode {
```
I'd be interested in knowing where I can get access to this node from? What API it is associated with?

hubertp replied:
Not sure I understand what you mean, but `DebuggerMessageHandler` was using the nested class, hence the extraction of the interface that can safely stay in `runtime` project.

JaroslavTulach committed:
Removing accidental change in ignore files

JaroslavTulach commented on engine/runtime-instrument-repl-debugger/src/main/java/org/enso/interpreter/instrument/ReplDebuggerInstrument.java:
```diff
@@ -85,7 +85,8 @@ protected OptionDescriptors getOptionDescriptors() {
 
   /** The actual node that's installed as a probe on any node the instrument was launched for. */
-  public static class ReplExecutionEventNode extends ExecutionEventNode {
+  public static class ReplExecutionEventNodeImpl extends ExecutionEventNode
```
Probably this class no longer needs to be `public` when it implements a public interface.

hubertp replied:
Sure, yeah

JaroslavTulach reviewed (commented):
Thank you very much Hubert for finishing this PR! At the end the changes to `build.sbt` look decent, small and logical, however I'd have a hard time trying to come up with them myself.

Should this _little restructuring_ of the family of `runtime-*` projects solve our problems with incremental compilation, then I wholeheartedly **approve**!

Btw. I cannot approve technically as I started the PR myself. I can just comment. However I certainly support integration of this PR.

hubertp committed:
IdExecutionService is optional

Lack of IdExecutionInstrument service would cause tests to blow up.
Where in reality, it is not necessary there.
Using Optional to avoid `if (x != null) ...` all over the place.

hubertp committed:
Move EPB to its own subproject

Minor cavet: added PrimArray to be able to pass arguments in
JsForeignNode, which is a stripped down version of Array in runtime.
Replacing `apply` with `call` and removing `Array` wrapping didn't do
the trick, unfortunately.

hubertp committed:
Remove FixInstrumentation workaround

hubertp committed:
DRY

hubertp committed:
Address PR review

hubertp commented:
@JaroslavTulach I'm done. WDYT?

hubertp committed:
nit

hubertp committed:
One less null check

hubertp commented:
Will remove myself from the review since I pretty much went medieval on it myself

JaroslavTulach commented on build.sbt:
```diff
@@ -260,6 +260,10 @@ lazy val enso = (project in file("."))
+    `runtime-instrument-id-execution`,
+    `runtime-instrument-repl-debugger`,
+    `runtime-instrument-runtime-server`,
+    `runtime-with-instruments`,
```
+1

JaroslavTulach commented on engine/runtime-language-epb/src/main/java/org/enso/interpreter/epb/node/JsForeignNode.java:
```diff
@@ -34,7 +33,7 @@ Object doExecute(
     try {
       return interopLibrary.invokeMember(
-          getForeignFunction(), "apply", arguments[0], new Array(positionalArgs));
+          getForeignFunction(), "apply", arguments[0], new PrimArray(positionalArgs));
```
OK.

JaroslavTulach reviewed (commented):
I continue to like the change and I support integration.

hubertp committed:
simplify

hubertp committed:
update licenses

hubertp committed:
nits

hubertp committed:
more cleanup (I promise, I'm not adding anything new anymore)

hubertp committed:
update documentation

hubertp committed:
Added some logging as per request

radeusgd committed:
Fix legal review

mergify[bot] committed:
Merge branch 'develop' into wip/engine/SeparateIdExecutionInstrument_182209126

radeusgd commented on build.sbt:
```diff
@@ -1197,13 +1225,9 @@ lazy val runtime = (project in file("engine/runtime"))
       "com.novocode"        % "junit-interface"       % "0.11"            % Test exclude ("junit", "junit-dep")
     ),
     // Note [Unmanaged Classpath]
-    Test / unmanagedClasspath += (baseDirectory.value / ".." / ".." / "app" / "gui" / "view" / "graph-editor" / "src" / "builtin" / "visualization" / "native" / "inc"),
```
If you move this line, I think you should also move the associated comment.

radeusgd replied:
Also btw. do you know why is that there?

hubertp replied:
Yeah, one of tests depends on the file shared between gui and engine

radeusgd replied:
I see, I still don't understand why we need to signalize this dependency. The test will just load this file at runtime, right? So I don't know why SBT must know about it (I'm not saying it doesn't have to, I just cannot see the reason).

hubertp replied:
how else would it know without hardcoding the path?

radeusgd replied:
Oh, no way around it.

My question is - why does it have to know?

hubertp replied:
Because of this single test:
```
[info] - should run internal IDE visualisation preprocessor catching error *** FAILED ***
[info]   java.io.FileNotFoundException: resource 'error_preprocessor.enso' was not found in the classpath from the given classloader.
[info]   at scala.io.Source$.fromResource(Source.scala:183)
```

Either runtime has to have it and then gui will have to know how to find it, or the other way around. I left it as is, i.e, the latter.

radeusgd commented on build.sbt:
```diff
@@ -1287,6 +1298,65 @@ lazy val runtime = (project in file("engine/runtime"))
+  )
+  .dependsOn(runtime)
+
+lazy val `runtime-with-instruments`  = (project in file("engine/runtime-instrument"))
```
I'd suggest making the project directory and project variable name consistent, unless we have a solid reason not to.

hubertp replied:
Thanks! That was on my list yesterday but I seemed to have forgotten about it

radeusgd commented on docs/infrastructure/sbt.md:
```diff
@@ -122,26 +120,17 @@ information is used by `enso --version`.
-stops the current compilation task, asking the user to restart it.
+this, so un-changed instruments get overwritten.
+
+In the past we had a pre-compile task (defined in
+[`FixInstrumentsGeneration`](../../project/FixInstrumentsGeneration.scala)) that
+detected changes to instruments and if only one of them was to be recompiled, it
```
This link is not valid anymore. I'd suggest removing the link.

hubertp replied:
Mostly left it for historical reasons, but sure, can remove it

radeusgd replied:
I think the description can stay. I'd just remove the link because it doesn't work anyway so it may be confusing (why the link doesn't work). You may change it to a GitHub link to some old commit on develop.

radeusgd commented on docs/runtime/instruments.md:
```diff
@@ -32,7 +31,6 @@ Unfortunately, when doing an incremental compilation, only the changed files are
-[`FixInstrumentsGeneration.scala`](../../project/FixInstrumentsGeneration.scala)
-task which detects changes to any of the instruments and forces recompilation of
-all instruments in the project by removing their classfiles.
+To fix that, individual services have to be placed in separate subprojects of
+`runtime` and later META-INF registrations are concatenated in the final uber
+jar.
```
What does it mean that project X is a 'subproject of runtime'? It doesn't sound right to me. Maybe:
```suggestion
To fix that, individual services have to be placed in separate subprojects depending on
`runtime` and aggregated under `runtime-with-instruments`. Later the META-INF registrations are concatenated in the final uber jar.
```

hubertp committed:
Revert "Added some logging as per request"

This reverts commit 11abd7c8269eea58ce65db8e651dca246fb2ad3d.

hubertp committed:
PR review

hubertp committed:
Update docs/runtime/instruments.md

Co-authored-by: Radosław Waśko <radoslaw.wasko@enso.org>

radeusgd commented on engine/runtime-language-epb/src/main/java/org/enso/interpreter/epb/node/JsForeignNode.java:
```diff
@@ -34,7 +33,7 @@ Object doExecute(
     try {
       return interopLibrary.invokeMember(
-          getForeignFunction(), "apply", arguments[0], new Array(positionalArgs));
+          getForeignFunction(), "apply", arguments[0], new ReadOnlyArray(positionalArgs));
```
What does this change imply?

radeusgd replied:
Ok, sorry, the `ReadOnlyArray` doc seems to explain it, I didn;t notice it before.

hubertp replied:
Functionally - nothing. We just cannot import Array here since `runtime-language-epb` is a dependency of `runtime` where it is defined.

radeusgd commented on docs/runtime/instruments.md:
```diff
@@ -32,7 +31,5 @@ Unfortunately, when doing an incremental compilation, only the changed files are
-[`FixInstrumentsGeneration.scala`](../../project/FixInstrumentsGeneration.scala)
-task which detects changes to any of the instruments and forces recompilation of
-all instruments in the project by removing their classfiles.
+To fix that, individual services have to be placed in separate subprojects depending on
+`runtime` and aggregated under `runtime-with-instruments`. Later the META-INF registrations are concatenated in the final uber jar.
```
Sorry, my GH suggestion was not properly formatted (as hard to do so in the Web UI) - please run `prettier` on it.

radeusgd commented on engine/runtime/src/main/java/org/enso/interpreter/Language.java:
```diff
@@ -108,12 +110,15 @@ protected Context createContext(Env env) {
         new Context(
             this, getLanguageHome(), env, notificationHandler, lockManager, distributionManager);
-    InstrumentInfo idValueListenerInstrument =
-        env.getInstruments().get(IdExecutionInstrument.INSTRUMENT_ID);
-    idExecutionInstrument = env.lookup(idValueListenerInstrument, IdExecutionInstrument.class);
+    idExecutionInstrument =
+        Optional.ofNullable(env.getInstruments().get(IdExecutionService.INSTRUMENT_ID))
+            .map(
+                idValueListenerInstrument ->
+                    env.lookup(idValueListenerInstrument, IdExecutionService.class));
```
Why do we allow `null` here? Shouldn't we throw an exception if the instrument is not available?

radeusgd replied:
Is it ever possible to not have the instrument loaded in any other scenario other than a sever misconfiguration?

hubertp replied:
That would make it impossible to setup unit tests in `runtime`. All tests that depend on the instrumentation have been moved to `runtime-with-instruments/src/test`

radeusgd replied:
Well the instruments should be already available in `runtime-with-instruments`, so I'm not sure if I understand the issue here. Is there a scenario when we run the runtime and the instruments are unavailable? In the tests it sounds like they totally should be available.

hubertp replied:
Yes, all tests that are in `runtime/src/test`. Those don't need instrumentation. That's why we moved (almost) all in the `instrument` package to `runtime-with-instruments` because those actually do require instrumentation.

hubertp committed:
prettify

radeusgd reviewed (approved):
Can't wait to see how the builds will behave now :tada:

hubertp committed:
Bring back printing hashes

hubertp committed:
Update package hash

hubertp committed:
Add logs to understand CI issues

mergify[bot] committed:
Merge branch 'develop' into wip/engine/SeparateIdExecutionInstrument_182209126

hubertp committed:
Skip internal artifacts in checksum

This is causing problems because version may be set differently for
different jobs and it is part of `module.toString`

hubertp committed:
update checksum

hubertp committed:
Revert "Add logs to understand CI issues"

This reverts commit c275b4a296623520ef39e966fba38f3a0dd2fa1e.
Issue was diagnosed and resolved.

mergify[bot] committed:
Merge branch 'develop' into wip/engine/SeparateIdExecutionInstrument_182209126
